chichiwang/exercise-js
# Convert the Express server to TypeScript

## Description
Linked to Issue: #18 
Convert the express server `server/index.mjs` to a TypeScript implementation. Ensure that linting/tests/scripts continue to work as intended.

### NOTE
This change breaks the npm script `start`, as the `nodemon` watcher watching the script `server/index.mjs` will fail. Given that we will need to rebuild the server on each change now, we will have to rework the watch process. Future changes, when React is introduced, will require a bundler (such as [Webpack](https://webpack.js.org/)) and we may want Hot Module Reloading as well. This will require some rework of the watch process.

## Changes
* devDependencies added to `package.json`:
  * `@babel/preset-typescript`
  * `@types/express`
  * `@types/node`
  * `@typescript-eslint/eslint-plugin`
  * `@typescript-eslint/parser`
  * `babel-jest`
  * `typescript`
* Scripts added to `package.json`:
  * `clean`: Remove the `dist/` build directory
  * `build:server`: Transpile the server with TypeScript
* Convert `server/index.mjs` to `server/index.ts`
* Update `server/__tests__/index.test.js to require the new `index.ts` file for testing
* Update `.eslintrc.js` to parse TypeScript and ignore the `dist/` directory
* Update `.gitignore` to exclude the `dist/` directory
* Update `babel.config.js` to use `@babel/preset-typescript` plugin
* Add a `tsconfig.json` to provide baseline TypeScript configurations for the project
* Add a `tsconfig.server.json` to target the server build specifically

## Steps to QA
* Pull down and switch to this branch
* Run `npm run lint` to ensure passing linter
* Run `npm run test` to ensure passing tests and full coverage
* Run `npm run build:server` and then run `node dist/index.js` to ensure the built server works correctly
  * Navigate to `http://127.0.0.1:8080` in the browser and ensure no errors
